### PR TITLE
Persistable Alarms and YoungGenTuningPolicy

### PR DESCRIPTION
**Pull request by sidheart**
Fixes #: N/A

Description of changes: 
Each of the following is split into a logical commit to make reviewing easier using the Github UI
- Make JvmActionsAlarmMonitor persistable 
- Add logic to detect and fix missized JVM heap generations

Tests: JvmGenerationTuningPolicyTest

If new tests are added, how long do the new ones take to complete: Negligible

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: src/main/java/com/amazon/opendistro/elasticsearch/performanceanalyzer/rca/framework/api/aggregators/PersistableSlidingWindow.java [hunk 1]
```diff
@@ -37,20 +38,25 @@
...
     super(slidingWindowSize, timeUnit);
     this.pathToFile = filePath;
-    try {
-      if (Files.exists(pathToFile)) {
-        load(pathToFile);
-      }
-    } catch (IOException e) {
-      LOG.error("Couldn't load SlidingWindow data from {}", pathToFile, e);
+    if (this.pathToFile == null) {
+      this.enablePersistence = false;
```
Comment (vigyasharma) [inline src/main/java/com/amazon/opendistro/elasticsearch/performanceanalyzer/rca/framework/api/aggregators/PersistableSlidingWindow.java:54]: Minor: I feel like if we are giving this flag then this shouldn't really be a derived class, we should just add this ability to the base SlidingWindow class (which may be a good idea for a separate PR). Choosing a `PersistableSlidingWindow` instead of a `SlidingWindow` implies that you want persistence.

File: src/main/java/com/amazon/opendistro/elasticsearch/performanceanalyzer/rca/framework/api/aggregators/PersistableSlidingWindow.java [hunk 2]
```diff
@@ -95,4 +127,24 @@ protected synchronized void write() throws IOException {
     // atomic rotate
     FileRotate.rotateFile(tmpPath, pathToFile);
   }
+
+  /**
+   * Attempts to write the window data to disk if enough time has passed since the last write
+   *
+   * <p>If this method tries and fails to write data to disk, it will print an error message but will
+   * not throw any exception, responsibility is on the caller
```
Comment (vigyasharma) [inline src/main/java/com/amazon/opendistro/elasticsearch/performanceanalyzer/rca/framework/api/aggregators/PersistableSlidingWindow.java:135]: Since *responsibility for error handling is on the caller*, we should just throw IOException here and let the caller log it or handle it as needed.

File: src/main/java/com/amazon/opendistro/elasticsearch/performanceanalyzer/rca/framework/api/aggregators/PersistableSlidingWindow.java [hunk 3]
```diff
@@ -72,6 +81,26 @@ protected synchronized void load(Path path) throws IOException {
     }
   }
 
+  /**
+   * Attempts to load data from the file on disk into the SlidingWindow
+   *
+   * <p>If this function fails to load the file it will just print an error log and the caller will
+   * be allowed to continue on. It is the caller's responsibility to ensure their data has been loaded
```
Comment (vigyasharma) [inline src/main/java/com/amazon/opendistro/elasticsearch/performanceanalyzer/rca/framework/api/aggregators/PersistableSlidingWindow.java:88]: > It is the caller's responsibility to ensure their data has been loaded 
Then let's throw IOException and let caller catch it, to ensure that they've handled this case. Otherwise how will caller know if load failed?

File: src/main/java/com/amazon/opendistro/elasticsearch/performanceanalyzer/rca/framework/api/aggregators/PersistableSlidingWindow.java [hunk 4]
```diff
@@ -72,6 +81,26 @@ protected synchronized void load(Path path) throws IOException {
...
+   * properly into the SlidingWindow.
+   */
+  protected void tryLoad() {
```
Comment (vigyasharma) [inline src/main/java/com/amazon/opendistro/elasticsearch/performanceanalyzer/rca/framework/api/aggregators/PersistableSlidingWindow.java:91]: If we remove the try-catch (let caller handle) and have `if (!enablePersistence)` check in `load()` then do we need this fn? 
same for `tryWrite()`

**Review by vigyasharma — CHANGES_REQUESTED** [review]

File: src/main/java/com/amazon/opendistro/elasticsearch/performanceanalyzer/rca/framework/api/aggregators/PersistableSlidingWindow.java (hunk 4 above)
Comment (sidheart) [reply src/main/java/com/amazon/opendistro/elasticsearch/performanceanalyzer/rca/framework/api/aggregators/PersistableSlidingWindow.java]: Removed both

File: src/main/java/com/amazon/opendistro/elasticsearch/performanceanalyzer/rca/framework/api/aggregators/PersistableSlidingWindow.java (hunk 2 above)
Comment (sidheart) [reply src/main/java/com/amazon/opendistro/elasticsearch/performanceanalyzer/rca/framework/api/aggregators/PersistableSlidingWindow.java]: Done

File: src/main/java/com/amazon/opendistro/elasticsearch/performanceanalyzer/rca/framework/api/aggregators/PersistableSlidingWindow.java (hunk 3 above)
Comment (sidheart) [reply src/main/java/com/amazon/opendistro/elasticsearch/performanceanalyzer/rca/framework/api/aggregators/PersistableSlidingWindow.java]: Done

File: src/main/java/com/amazon/opendistro/elasticsearch/performanceanalyzer/rca/framework/api/aggregators/PersistableSlidingWindow.java (hunk 1 above)
Comment (sidheart) [reply src/main/java/com/amazon/opendistro/elasticsearch/performanceanalyzer/rca/framework/api/aggregators/PersistableSlidingWindow.java:53]: You're right, I think the enablePersistence design is unnecessarily complicated to allow the caller to pass in a null path. 

The reason I built it this way was to avoid modifying all callers. We can and should change it in a future PR.

File: src/main/java/com/amazon/opendistro/elasticsearch/performanceanalyzer/decisionmaker/deciders/jvm/JvmGenerationTuningPolicy.java [hunk 5]
```diff
@@ -0,0 +1,216 @@
...
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+
+/**
+ * JvmGenerationTuningPolicy decides if the JVM heap generations could be resized to improve application
+ * performance and suggests actions to take to achieve improved performance.
+ */
+public class JvmGenerationTuningPolicy implements DecisionPolicy {
+  private static final Logger LOG = LogManager.getLogger(JvmGenerationTuningPolicy.class);
+  private static final long COOLOFF_PERIOD_IN_MILLIS = 48L * 24L * 60L * 60L * 1000L;
```
[the remark this answers is not in the document]
Comment (sidheart) [reply src/main/java/com/amazon/opendistro/elasticsearch/performanceanalyzer/decisionmaker/deciders/jvm/JvmGenerationTuningPolicy.java]: THANK YOU

File: src/main/java/com/amazon/opendistro/elasticsearch/performanceanalyzer/decisionmaker/deciders/jvm/JvmActionsAlarmMonitor.java [hunk 6]
```diff
@@ -48,14 +69,14 @@ public void recordIssue(long timeStamp, double value) {
...
-    if (dayMonitor.size() >= dayBreachThreshold) {
+    if (dayMonitor.readSum() >= dayBreachThreshold) {
       weekMonitor.next(new SlidingWindowData(dataPoint.getTimeStamp(), dataPoint.getValue()));
     }
   }
 
   private void evaluateAlarm() {
     if (alarmHealthy) {
-      if (weekMonitor.size() >= weekBreachThreshold) {
+      if (weekMonitor.readSum() >= weekBreachThreshold) {
```
[the remark this answers is not in the document]
Comment (sidheart) [reply src/main/java/com/amazon/opendistro/elasticsearch/performanceanalyzer/decisionmaker/deciders/jvm/JvmActionsAlarmMonitor.java]: Reverted

File: src/main/java/com/amazon/opendistro/elasticsearch/performanceanalyzer/decisionmaker/deciders/jvm/JvmActionsAlarmMonitor.java [hunk 7]
```diff
@@ -48,14 +69,14 @@ public void recordIssue(long timeStamp, double value) {
     SlidingWindowData dataPoint = new SlidingWindowData(timeStamp, value);
     dayMonitor.next(dataPoint);
     // If we've breached the day threshold, record it as a bad day this week.
-    if (dayMonitor.size() >= dayBreachThreshold) {
+    if (dayMonitor.readSum() >= dayBreachThreshold) {
```
[the remark this answers is not in the document]
Comment (sidheart) [reply src/main/java/com/amazon/opendistro/elasticsearch/performanceanalyzer/decisionmaker/deciders/jvm/JvmActionsAlarmMonitor.java]: Reverted

File: src/main/java/com/amazon/opendistro/elasticsearch/performanceanalyzer/decisionmaker/deciders/jvm/JvmActionsAlarmMonitor.java [hunk 8]
```diff
@@ -19,26 +19,47 @@
...
+    Path dayMonitorPath = null;
+    Path weekMonitorPath = null;
+    if (persistencePath != null) {
+      String persistenceBase = persistencePath.getParent().toString();
+      String persistenceFile = persistencePath.getFileName().toString();
+      dayMonitorPath = Paths.get(persistenceBase, DAY_PREFIX + persistenceFile);
+      weekMonitorPath = Paths.get(persistenceBase, WEEK_PREFIX + persistenceFile);
+    }
+    dayMonitor = new BucketizedSlidingWindow((int) TimeUnit.DAYS.toMinutes(1), 30, TimeUnit.MINUTES, dayMonitorPath);
+    weekMonitor = new BucketizedSlidingWindow(4, 1, TimeUnit.DAYS, weekMonitorPath);
```
[the remark this answers is not in the document]
Comment (sidheart) [reply src/main/java/com/amazon/opendistro/elasticsearch/performanceanalyzer/decisionmaker/deciders/jvm/JvmActionsAlarmMonitor.java:53]: Yes, if it's null we won't attempt to do any persistence

File: src/main/java/com/amazon/opendistro/elasticsearch/performanceanalyzer/decisionmaker/deciders/configs/jvm/JvmGenerationTuningPolicyConfig.java [hunk 9]
```diff
@@ -0,0 +1,67 @@
...
+ *    "jvm-generation-tuning-policy-config": {
+ *       "enabled": true,
+ *       "should-decrease-young-gen": false,
+ *       ...
+ *    }
+ * }
+ */
+public class JvmGenerationTuningPolicyConfig {
+  private static final String ENABLED = "enabled";
+  private static final String SHOULD_DECREASE_YOUNG_GEN = "should-decrease-young-gen";
```
[the remark this answers is not in the document]
Comment (sidheart) [reply src/main/java/com/amazon/opendistro/elasticsearch/performanceanalyzer/decisionmaker/deciders/configs/jvm/JvmGenerationTuningPolicyConfig.java]: Done

File: src/main/java/com/amazon/opendistro/elasticsearch/performanceanalyzer/decisionmaker/deciders/configs/DeciderConfig.java [hunk 10]
```diff
@@ -43,10 +44,12 @@
     private static final String CACHE_CONFIG_NAME = "cache-type";
     private static final String WORKLOAD_CONFIG_NAME = "workload-type";
     private static final String OLD_GEN_DECISION_POLICY_CONFIG_NAME = "old-gen-decision-policy-config";
+    private static final String JVM_GENERATION_TUNING_POLICY_CONFIG_NAME = "jvm-generation-tuning-policy-config";
```
[the remark this answers is not in the document]
Comment (sidheart) [reply src/main/java/com/amazon/opendistro/elasticsearch/performanceanalyzer/decisionmaker/deciders/configs/DeciderConfig.java]: s/generation/gen across many files

File: src/main/java/com/amazon/opendistro/elasticsearch/performanceanalyzer/decisionmaker/actions/JvmGenerationAction.java [hunk 11]
```diff
@@ -0,0 +1,99 @@
...
+  public Map<NodeKey, ImpactVector> impact() {
+    Map<NodeKey, ImpactVector> impact = new HashMap<>();
+    for (NodeKey key : impactedNodes()) {
+      impact.put(key, NO_IMPACT);
+    }
+    return impact;
+  }
+
+  @Override
+  public String summary() {
```
[the remark this answers is not in the document]
Comment (sidheart) [reply src/main/java/com/amazon/opendistro/elasticsearch/performanceanalyzer/decisionmaker/actions/JvmGenerationAction.java]: Done

File: src/main/java/com/amazon/opendistro/elasticsearch/performanceanalyzer/decisionmaker/deciders/configs/jvm/JvmGenTuningPolicyConfig.java [hunk 12]
```diff
@@ -0,0 +1,66 @@
...
+
+/**
+ * Configures various options for the
+ * {@link JvmGenTuningPolicy}
+ *
+ * <p>The config follows the format below
+ *  "decider-config-settings": {
+ *    "jvm-generation-tuning-policy-config": {
+ *       "enabled": true,
+ *       "should-decrease-young-gen": false,
```
Comment (vigyasharma) [inline src/main/java/com/amazon/opendistro/elasticsearch/performanceanalyzer/decisionmaker/deciders/configs/jvm/JvmGenTuningPolicyConfig.java:30]: nit: update here as well

**Review by vigyasharma — APPROVED** [review]

File: src/main/java/com/amazon/opendistro/elasticsearch/performanceanalyzer/decisionmaker/deciders/configs/jvm/JvmGenTuningPolicyConfig.java (hunk 12 above)
Comment (sidheart) [reply src/main/java/com/amazon/opendistro/elasticsearch/performanceanalyzer/decisionmaker/deciders/configs/jvm/JvmGenTuningPolicyConfig.java]: Done

**Review by vigyasharma — DISMISSED** [review]

File: src/main/java/com/amazon/opendistro/elasticsearch/performanceanalyzer/rca/framework/api/aggregators/BucketizedSlidingWindow.java [hunk 13]
```diff
@@ -41,6 +62,11 @@ public void next(SlidingWindowData e) {
         firstElement.value += e.getValue();
         add(e);
         pruneExpiredEntries(e.getTimeStamp());
+        try {
+          write(); // Try to persist the data whenever we complete writing a bucket
+        } catch (IOException ex) {
```
Comment (ktkrg) [inline src/main/java/com/amazon/opendistro/elasticsearch/performanceanalyzer/rca/framework/api/aggregators/BucketizedSlidingWindow.java:67]: Can we add a metric here tracking persistence failure?

File: src/main/java/com/amazon/opendistro/elasticsearch/performanceanalyzer/rca/framework/api/aggregators/PersistableSlidingWindow.java [hunk 14]
```diff
@@ -37,20 +37,29 @@
...
+    if (this.pathToFile == null) {
+      this.enablePersistence = false;
+    } else {
+      this.enablePersistence = true;
+      // setting the last write time to now will cause our first write to occur 5 minutes after construction
+      this.lastWriteTimeEpochMs = Instant.now().toEpochMilli();
+      try {
+        load(this.pathToFile);
+      } catch (IOException ex) {
+        LOG.error("Unable to load previous data from {} into {}", this.pathToFile, getClass().getSimpleName());
```
Comment (ktkrg) [inline src/main/java/com/amazon/opendistro/elasticsearch/performanceanalyzer/rca/framework/api/aggregators/PersistableSlidingWindow.java:61]: Same, track load failure with a metric?

**Review by ktkrg — COMMENTED** [review]
LGTM, left a couple of comments related to adding operational metrics.

File: src/main/java/com/amazon/opendistro/elasticsearch/performanceanalyzer/rca/framework/api/aggregators/BucketizedSlidingWindow.java (hunk 13 above)
Comment (sidheart) [reply src/main/java/com/amazon/opendistro/elasticsearch/performanceanalyzer/rca/framework/api/aggregators/BucketizedSlidingWindow.java:71]: Done

**Review by vigyasharma — APPROVED** [review]

**Review by vigyasharma — APPROVED** [review]

**Comment by sidheart** [comment]
Superseded by #468
